Nextclade Subworkflow

ammaraziz opened the pull request:
Implement nextclade on a per organism basis (not per sample).  One issue with this PR is testing, should subworkflows require separate testing?

Please let me know what you think and if you suggest any changes.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [X] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/artic-network/amplicon-nf/tree/master/.github/CONTRIBUTING.md)
- [X] Make sure your code lints (`nf-core pipelines lint`). 
WARNINGS are produced for many of the pipeline files - For the new subworkflow warning is related to outputting a version, which does happen.
- [X] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [X] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
Haven't added anything to change log - it's not up to date?
- [X] `README.md` is updated (including new tool citations and authors/contributors).

ammaraziz committed:
feat: add nextclade

ammaraziz committed:
chore: new flag to rename header or not

ammaraziz committed:
chore: add missing '

ammaraziz committed:
chore: fix indent

ammaraziz committed:
chore: add fasta_header_extra

ammaraziz committed:
chore: replace single with double quotes

ammaraziz committed:
chore: replace single with double quotes

ammaraziz committed:
chore: testing parsing

ammaraziz committed:
chore: testing parsing

ammaraziz committed:
chore: testing parsing

ammaraziz committed:
chore: testing parsing

ammaraziz committed:
chore: testing parsing

ammaraziz committed:
chore: basic implementation of nextclade in run report

ammaraziz committed:
chore: config for seqkit replace nextclade

ammaraziz committed:
chore: nextclade workflow

ammaraziz committed:
chore: renamed workflow

ammaraziz committed:
chore: removed nextclade from workflow to simplify

ammaraziz committed:
chore: removed unused import

ammaraziz committed:
chore: simplify nextclade params

ammaraziz committed:
chore: simplify nextclade params

ammaraziz committed:
docs: added docs for nextclade module

ammaraziz committed:
Merge branch 'dev' into dev-nextclade

ammaraziz committed:
docs: added info on nextclade

ammaraziz committed:
Merge branch 'dev-nextclade' of github.com:ammaraziz/amplicon-nf into dev-nextclade

BioWilko commented on .github/workflows/linting.yml:
```diff
@@ -1,80 +0,0 @@
-name: nf-core linting
```
Please don't delete my actions

ammaraziz replied:
Sorry!

BioWilko commented on modules/local/generate_run_report/templates/generate_run_report.py:
```diff
@@ -856,6 +856,23 @@ def primer_mismatch_heatmap(
+                    "Cov": row.get("qc_overall_score", ""),
+                }
+            )
+payload["nextclade_table"] = nextclade_rows
```
as far as I can tell this doesn't actually make it into the report itself (no changes to the template), ideally any new outputs should be included in a tab in the run report.

BioWilko commented on workflows/amplicon-nf.nf:
```diff
@@ -3,8 +3,8 @@
     IMPORT MODULES / SUBWORKFLOWS / FUNCTIONS
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 */
```
Please remove these changes, I assume they were accidental

BioWilko reviewed (changes requested):
Seems like a good start for nextclade support but I think it needs to be integrated into the pipeline a bit better before we can merge, specifically I think nextclade results need to be nicely included in the run / sample reports when run.

The docs also need to make it very clear that the nextclade functionality doesn't support multi virus runs / multi pathogen schemes

ammaraziz committed:
chore: correct merge error; changed nf-schema to correct version

ammaraziz committed:
chore: restore

ammaraziz committed:
Dev nextclade html (#11)

Revamp nextclade subworkflow, changes

    HTML output
    Better handling of optional inputs for GENERATE_RUN_REPORT
    Moved nextclade into amplicon-nf
    Capture nextclade
    Removed up unused params



* chore: format imports correctly

* chore: remove channel for nextclade tag

* chore: rename module

* chore: rename subworkflow

* add lineage tab into the sample run report

* chore: first attempt implement html output for nextclade

* chore: add nextclade table to template

* chore: add nextclade for file renaming

* chore: rename nextclade tags to be simple

* chore: simplify nextclade data parsing into payload

* chore: WORK IN PROGRESS - adding optional nextclade channel

* chore: move nextclade workflow into amplicon-nf.nf

* add nextclade tsv to inputs

* chore: simplify code

* chore: first working version of new nextclade workflow

* chore: added NO_FILE to handle optional input

* feat: nextclade html output

* chore: simplify run report code

* chore: remove unused param

* chore: remove nextclade settings; tsv output not produced

* chore: remove fasta_header_extra param

* chore: output software version

* chore: capture nextclade version

---------

Co-authored-by: Josh Zhang <zjs0317@gmail.com>

BioWilko commented on workflows/amplicon-nf.nf:
```diff
@@ -315,36 +325,32 @@ workflow AMPLICON_NF {
-    }
+    // massage optional inputs
+    ch_msas_opt       = params.primer_mismatch_plot ? ch_msas_by_scheme : channel.empty()
+    ch_nextclade_opt  = params.nextclade ? ch_nextclade_tsv :  channel.fromPath("$projectDir/assets/NO_FILE")
```
```suggestion
    ch_nextclade_opt  = params.nextclade ? ch_nextclade_tsv :  channel.empty()
```

BioWilko commented on workflows/amplicon-nf.nf:
```diff
@@ -315,36 +325,32 @@ workflow AMPLICON_NF {
+                depth,
+                amp,
+                cov,
+                msas ?: [],
```
pack the nextclade TSV in here

ammaraziz replied:
Unfortunately this doesn't work because the `meta` object which is attached to each sample/item running through the workflow is stripped out during the  nextclade subworkflow.

ammaraziz replied:
Actually the above code still produces an error, I think due to passing a raw channel. I went down a rabbit hole and I decided to simplify the Nextclade code as it was badly written. It's now very similar to the rest of the amplicon-nf code base.

ammaraziz committed:
clean up and simplify implementation of nextclade subworkflow

ammaraziz committed:
remove unused config for nextclade, no longer using seqkit

ammaraziz committed:
remove dup code

ammaraziz committed:
changed to use nextclade dataset module as to capture the nextclade dataset version

ammaraziz committed:
chore: remove view

ammaraziz committed:
chore: removed - unused

ammaraziz committed:
Implement local nextclade dataset inputs  (#12)

* chore: remove view

* chore: removed - unused

* feat: implement offline datasets mode

ammaraziz committed:
Merge branch 'dev-nextclade' of github.com:ammaraziz/amplicon-nf-dev1 into dev-nextclade

ammaraziz committed:
chore: update nextclade_getdataset module - tag saved to versions

ammaraziz committed:
chore: update cli help docs for nextclade

ammaraziz committed:
chore: publish nextclade dataset directory when downloaded

ammaraziz committed:
chore: publish nextclade results

ammaraziz committed:
Dev nextclade testing (#14)

* tests: add nextclade params

* chore: fix bug

* add nextclade outputs; update snapshot

ammaraziz committed:
Merge branch 'dev' into dev-nextclade

ammaraziz committed:
Add nextclade to docs (#15)

* docs: expand nextclade section

* docs: add nextclade info

ammaraziz committed:
chore: fix bug in nextclade html output.

Nextclade outputs good, mediocre and bad. The code was incorrectly checking for fail instead of bad. This fixes that bug

ammaraziz committed:
Merge branch 'dev-nextclade' of github.com:ammaraziz/amplicon-nf-dev1 into dev-nextclade

ammaraziz committed:
chore: add sorting to nextclade table

ammaraziz committed:
chore: remove format on nextclade tag

BioWilko committed:
Add concatenated FASTA output (#36)

* add multiconsensus FASTA output

* add resource limit overrides to the other profiles to compensate for standard always being applied

* remove resource limits from other profiles

* bump version string

* Add maxforks to get scheme to avoid race condition

* Add max retries to get_scheme instead of max forks

* update output doc

BioWilko committed:
Hotfix for epi2me compatibility

BioWilko committed:
Update test snapshot

BioWilko committed:
Low resource changes (#39)

* low resource / error handling improvements

* patch process high too

* apply requests to all labels

BioWilko committed:
Patch to fix nextflow 23.04.2 compatibility

BioWilko committed:
0.7.4 dev (#40)

* epi2me compatibility patch

* update config again

* remove newline

* adjust requests again

* and again

* Update snapshot

* don't store store in repo anymore

ammaraziz committed:
Merge branch 'dev-main' into dev-nextclade

BioWilko commented on conf/modules.config:
```diff
@@ -204,4 +197,20 @@ process {
+            mode: params.publish_dir_mode,
+            saveAs: { filename -> filename.equals('versions.yml') ? null : filename }
+        ]
+    }
```
Do we want to publish the nextclade dataset?

ammaraziz replied:
Yeah I thought it might be useful to publish the dataset so that it's only downloaded once and can be reused. But fine to remove if you think it's not useful.

BioWilko committed:
Merge branch 'main' into dev-nextclade

BioWilko reviewed (commented):
One thing I'm not sure about, otherwise I'm going to make some changes and push directly.

BioWilko committed:
Slightly opinionated updates :)

BioWilko committed:
Add nextclade JSON output, update docs

BioWilko reviewed (approved):
Okay, LGTM, how about you @ammaraziz ?

BioWilko committed:
Update nextclade part of usage doc

BioWilko committed:
Rephrase nextclade usage warning
